### PR TITLE
Ensure test run failures crash worker

### PR DESCRIPTION
**Pull request by novemberborn**
The first commit is there to catch more errors. The second commit ensures errors like https://github.com/avajs/ava/issues/1263 are surfaced without the test hanging.

### CODE REVIEW

**Comment by novemberborn** [comment]
Added tests to last commit. Let's hope they pass in CI 😄

File: lib/main.js [hunk 1]
```diff
@@ -70,7 +70,23 @@ globals.setImmediate(() => {
...
+			.then(exit)
+			.catch(err => {
+				let exception;
+				try {
+					exception = serializeError(err);
+				} catch (_) {
+					// Avoid using serializeError
+					const err = new Error('Runner failed with an unserializable exception');
+					exception = {
+						name: 'Error',
```
Comment (sindresorhus) [inline lib/main.js]: Super nitpick, but would prefer `err.name` here just for consistency with the other props and more future proof if we decide to change the error type or something.

File: lib/main.js [hunk 2]
```diff
@@ -70,7 +70,23 @@ globals.setImmediate(() => {
...
-		runner.run(options).then(exit);
+		runner.run(options)
+			.then(exit)
+			.catch(err => {
+				let exception;
+				try {
+					exception = serializeError(err);
+				} catch (_) {
+					// Avoid using serializeError
+					const err = new Error('Runner failed with an unserializable exception');
```
Comment (sindresorhus) [inline lib/main.js]: I wonder, should we put this logic in `serializeError` so it handles reporting about unserializable exceptions everywhere?

File: test/fixture/trigger-worker-exception/test-fallback.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import test from '../../../';
+
+test(() => {
+	return Promise.reject(new Error('Hi :)'));
```
Comment (sindresorhus) [inline test/fixture/trigger-worker-exception/test-fallback.js]: Would look better to use async function here and throw instead. Same with the below fixture.

File: lib/main.js (hunk 2 above)
Comment (novemberborn) [inline lib/main.js]: It's only called in three places. I'd prefer letting the error propagate correctly, and then treating it as an uncaught exception.

That said I've changed this code to `process.emit('uncaughtException', err)` and moved the fallback stuff to the `uncaughtException` handler in `test-worker.js`.

**Pull request merged**